ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/CheckpointHeaderFetcher.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.eth.sync;
+
+import tech.pegasys.pantheon.ethereum.core.BlockHeader;
+import tech.pegasys.pantheon.ethereum.eth.manager.EthContext;
+import tech.pegasys.pantheon.ethereum.eth.manager.EthPeer;
+import tech.pegasys.pantheon.ethereum.eth.manager.task.AbstractPeerTask.PeerTaskResult;
+import tech.pegasys.pantheon.ethereum.eth.manager.task.GetHeadersFromPeerByHashTask;
+import tech.pegasys.pantheon.ethereum.mainnet.ProtocolSchedule;
+import tech.pegasys.pantheon.metrics.MetricsSystem;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.UnaryOperator;
+
+public class CheckpointHeaderFetcher {
+
+  private final SynchronizerConfiguration syncConfig;
+  private final ProtocolSchedule<?> protocolSchedule;
+  private final EthContext ethContext;
+  private final UnaryOperator<List<BlockHeader>> checkpointFilter;
+  private final MetricsSystem metricsSystem;
+
+  public CheckpointHeaderFetcher(
+      final SynchronizerConfiguration syncConfig,
+      final ProtocolSchedule<?> protocolSchedule,
+      final EthContext ethContext,
+      final UnaryOperator<List<BlockHeader>> checkpointFilter,
```

Comment:
(optional) You could set this as a `Supplier<OptionalLong>` that provides the max block number to request, then we can avoid going over the limit for the last fast sync request.

Reply:
I'm going to leave this for now but have a note to revisit.  We need to ensure the last checkpoint header for fast sync is the pivot block and it sometimes needs to be explicitly inserted.  We could use an `Optional<BlockHeader> finalCheckpoint` and include all the logic inside `CheckpointHeaderFetcher` but then it winds up with a bunch of fast sync specific logic mixed in.  Need to play around with it but want to get something actually working first.